core/src/apps/bitcoin/sign_tx/payment_request.py

```diff
@@ -0,0 +1,88 @@
+from typing import TYPE_CHECKING
+
+from storage import cache
+from trezor import wire
+from trezor.crypto.curve import secp256k1
+from trezor.crypto.hashlib import sha256
+from trezor.enums import MemoType
+from trezor.utils import HashWriter
+
+from apps.common import coininfo
+from apps.common.address_mac import check_address_mac
+from apps.common.keychain import Keychain
+
+from .. import writers
+
+if TYPE_CHECKING:
+    from trezor.messages import TxAckPaymentRequest, TxOutput
+
+
+class PaymentRequestVerifier:
+    if __debug__:
+        # secp256k1 public key of m/0h for "all all ... all" seed.
+        PUBLIC_KEY = b"\x03\x0f\xdf^(\x9bZ\xefSb\x90\x95:\xe8\x1c\xe6\x0e\x84\x1f\xf9V\xf3f\xac\x12?\xa6\x9d\xb3\xc7\x9f!\xb0"
+    else:
+        PUBLIC_KEY = b""
+
+    def __init__(
+        self, msg: TxAckPaymentRequest, coin: coininfo.CoinInfo, keychain: Keychain
+    ) -> None:
+        self.h_outputs = HashWriter(sha256())
+        self.amount = 0
+        self.expected_amount = msg.amount
+        self.signature = msg.signature
+        self.h_pr = HashWriter(sha256())
+
+        if msg.nonce:
+            nonce = bytes(msg.nonce)
+            if cache.get(cache.APP_COMMON_NONCE) != nonce:
+                raise wire.DataError("Invalid nonce in payment request.")
+            cache.delete(cache.APP_COMMON_NONCE)
+        else:
+            nonce = b""
+            if msg.memos:
+                wire.DataError("Missing nonce in payment request.")
+
+        writers.write_bytes_fixed(self.h_pr, b"SL\x00\x24", 4)
+        writers.write_bytes_prefixed(self.h_pr, nonce)
+        writers.write_bytes_prefixed(self.h_pr, msg.recipient_name.encode())
+        writers.write_compact_size(self.h_pr, len(msg.memos))
+        for memo in msg.memos:
+            writers.write_uint32(self.h_pr, memo.type)
+            if memo.type == MemoType.COIN_PURCHASE:
+                assert memo.amount is not None  # checked by sanitizer
+                assert memo.coin_name is not None  # checked by sanitizer
+                assert memo.mac is not None  # checked by sanitizer
+                memo_coin = coininfo.by_name(memo.coin_name)
```

Comment:
This still only works for Bitcoin-like coins. I don't think we have, or even _can_ have, a general method of fetching the coin symbol given the SLIP44 identifier. In particular, such method would fail for ERC20 tokens.

ISTM we could simply include the symbol in the payment request? We have no way to identify the amount anyway, so the only thing the use of SLIP44 serves (if i understand right) is cross-verification that the address matches the coin -- but, as pointed out, this can't really work.

Not sure about the SLIP-24 text and the address MAC calculation either. Adding SLIP-44 is really weak.
Perhaps adding the full derivation path instead?

Reply:
The SLIP44 coin type serves to verify that the address matches the coin, e.g. ETH, not necessarily the token. As discussed, we will specify the token/symbol using a string. I think we should specify the amount itself as a string, because we might have a problem determining the number of decimal places the coin/token uses. See https://github.com/trezor/trezor-firmware/pull/2069/commits/1c5fb4284e8a34e69362c92e714572bd7b9f1f2b#diff-4568999ce3d84d878396d98f640d2cab8ee735bc568324e7e387ef4254c840a5R448-R452.